File: +bsv/fetchConnectivityImages.m
function status = fetchConnectivityImages(experimentID, saveFilePath)
% density.raw is in AP x DV x ML, 100um resolution
status = true;

% Create directory if it doesn't exist
if ~exist(saveFilePath, 'dir')
    [mkdirStatus, msg] = mkdir(saveFilePath);
    if ~mkdirStatus
        status = false;
        warning('Failed to create directory %s: %s', saveFilePath, msg);
        return;
    end
end

% Verify directory exists and is writable
if ~exist(saveFilePath, 'dir')
    status = false;
    warning('Directory %s does not exist after creation attempt', saveFilePath);
    return;
end

% Additional check - try to create a test file to verify write permissions
testFile = fullfile(saveFilePath, 'test_write.txt');
try
    fid = fopen(testFile, 'w');
    if fid == -1
        status = false;
        warning('Cannot write to directory %s - check permissions', saveFilePath);
        return;
    end
    fclose(fid);
    delete(testFile);
catch
    status = false;
    warning('Cannot write to directory %s - check permissions', saveFilePath);
    return;
end

zipFile = [saveFilePath, filesep, 'temp.zip'];

% Try to download the file with error handling
try
    % First try urlwrite
    urlwrite(sprintf('http://api.brain-map.org/grid_data/download/%d?include=density', experimentID), zipFile);
catch ME
    % If urlwrite fails, try websave as a fallback (newer MATLAB versions)
    try
        websave(zipFile, sprintf('http://api.brain-map.org/grid_data/download/%d?include=density', experimentID));
    catch ME2
        status = false;
        warning('Failed to download data for experiment %d: %s', experimentID, ME.message);
        return;
    end
end

% Unzip the file
try
    unzip(zipFile, saveFilePath);
catch ME
    status = false;
    warning('Failed to unzip data for experiment %d: %s', experimentID, ME.message);
    return;
end

filePath = [saveFilePath, '/density.raw'];
if ~exist(filePath, 'file')
    status = false;
    warning('Error downloading experiment %s, skipping.', experimentID);
end

end

File: bsv/fetch_connectivity_images.py
import os
import zipfile
import tempfile
import requests

PROJECTION_GRID_SIZE = (132, 80, 114)


def fetch_connectivity_images(experiment_id, save_file_path):
    """Download density.raw for an experiment. Returns True on success."""
    os.makedirs(save_file_path, exist_ok=True)

    raw_path = os.path.join(save_file_path, 'density.raw')
    if os.path.exists(raw_path):
        return True

    url = f'http://api.brain-map.org/grid_data/download/{experiment_id}?include=density'

    try:
        resp = requests.get(url, stream=True)
        resp.raise_for_status()
    except Exception as e:
        print(f'Warning: Failed to download data for experiment {experiment_id}: {e}')
        return False

    # Save zip and extract
    zip_path = os.path.join(save_file_path, 'temp.zip')
    try:
        with open(zip_path, 'wb') as f:
            for chunk in resp.iter_content(chunk_size=8192):
                f.write(chunk)
        with zipfile.ZipFile(zip_path, 'r') as z:
            z.extractall(save_file_path)
    except Exception as e:
        print(f'Warning: Failed to extract data for experiment {experiment_id}: {e}')
        return False
    finally:
        if os.path.exists(zip_path):
            os.remove(zip_path)

    if not os.path.exists(raw_path):
        print(f'Warning: density.raw not found after extraction for experiment {experiment_id}')
        return False

    return True
